ECMAScript Module Exports

awmichel opened the pull request:
Fixes #12 

See issue #12 for the reasoning behind the change. The change itself is to the import/export syntax of the `.mjs` files output by the `typescript_proto_library` rule. Previously they were CommonJS format which required tricking Rollup's CommonJS plugin into parsing `.mjs` files. Now the output files can be correctly parsed by Rollup without additional configuration.

Additionally, I added a Jasmine test spec to ensure the Rollup bundle has the correct outputs and functionality.

Dig-Doug commented on src/change_import_style.ts:
```diff
@@ -62,7 +62,7 @@ function convertToUmd(args: any, initialContents: string): string {
-function processCommonJs(args: any, initialContents: string): string {
+function convertToModule(args: any, initialContents: string): string {
   // Rollup can't resolve the commonjs exports when using goog.object.extend so we replace it with:
   // 'exports.MyProto = proto.namespace.MyProto;'
```
Can you update this comment?

Dig-Doug commented on src/change_import_style.ts:
```diff
@@ -62,7 +62,7 @@ function convertToUmd(args: any, initialContents: string): string {
-function processCommonJs(args: any, initialContents: string): string {
+function convertToModule(args: any, initialContents: string): string {
   // Rollup can't resolve the commonjs exports when using goog.object.extend so we replace it with:
   // 'exports.MyProto = proto.namespace.MyProto;'
```
Can you update this comment?

awmichel replied:
No problem, updated. I simplified it by noting the CommonJS -> ESM goal and provided a reference link to MDN.

Dig-Doug commented on src/change_import_style.ts:
```diff
@@ -74,15 +74,20 @@ function processCommonJs(args: any, initialContents: string): string {
 
     const exportSymbols = symbols.reduce((currentSymbols, symbol) => {
-      return currentSymbols + `exports.${symbol[0]} = ${symbol[1]};\n`;
+      return currentSymbols + `export const ${symbol[0]} = ${symbol[1]};\n`;
```
Instead of using `export const ...`, can we use `export { X as Y};`? It avoids any possible problems if the symbol was already defined as a variable:

```js
const Proto = 0;

export const Proto = Proto;
```

The `export-as` syntax also seems to fit this use case better since we're renaming things, not defining them.

https://developer.mozilla.org/en-US/docs/web/javascript/reference/statements/export

awmichel replied:
I attempted to use the `export-as` syntax, but wasn't able to make it work due to the way the protos are defined in a nested structure.

This example throws a syntax error:
```javascript
export { proto.test.bazel.proto.Pizza as Pizza };
```

There are a couple ways we could probably make it work if we needed to. Is the example you provided something you've encountered? Short of an export being named `goog` or `global`, I don't see a possibility for name conflicts.

Nevertheless, while experimenting with `export-as` I managed the improve the existing export to:
```javascript
export const { Pizza, PizzaSize } = proto.test.bazel.proto
```

awmichel commented:
@Dig-Doug I fixed the issue with the GRPC service definitions and added some additional tests to make sure the service definition and client worked as expected.

I encountered a bug in grpc-web that required some additional Rollup config (see: https://github.com/improbable-eng/grpc-web/issues/369), but it shouldn't take long to resolve that issue and remove the config exception from the tests.

awmichel commented:
@Dig-Doug I added a check to make sure we don't try to export nested structures, should catch any similar bugs to the case you provided. Also included that in the tests.
